redbot/cogs/mutes/mutes.py

```diff
@@ -0,0 +1,1212 @@
+                    to_del.append(user_id)
+                    try:
+                        log.debug("creating case")
+                        await modlog.create_case(
+                            self.bot,
+                            after.guild,
+                            datetime.utcnow(),
+                            "cunmute",
+                            user,
+                            None,
+                            _("Manually removed channel overwrites"),
+                            until=None,
+                            channel=after,
+                        )
+                        log.debug("created case")
+                    except RuntimeError as e:
+                        log.error(_("Error creating modlog case"), exc_info=e)
+            for u_id in to_del:
+                del self._channel_mutes[after.id][u_id]
+            await self.config.channel(after).muted_users.set(self._channel_mutes[after.id])
+
+    @commands.Cog.listener()
+    async def on_member_remove(self, member: discord.Member):
+        guild = member.guild
+        until = None
+        if guild.id in self._server_mutes:
+            if member.id in self._server_mutes[guild.id]:
+                until = self._server_mutes[guild.id][member.id]["until"]
+                del self._server_mutes[guild.id][member.id]
+        for channel in guild.channels:
+            if channel.id in self._channel_mutes:
+                if member.id in self._channel_mutes[channel.id]:
+                    del self._channel_mutes[channel.id][member.id]
+        mute_role = await self.config.guild(guild).mute_role()
+        if not mute_role:
+            return
+        if mute_role in [r.id for r in member.roles]:
+            async with self.config.guild(guild).removed_users() as removed_users:
+                removed_users[str(member.id)] = int(until) if until else None
+
+    @commands.Cog.listener()
+    async def on_member_join(self, member: discord.Member):
+        guild = member.guild
+        mute_role = await self.config.guild(guild).mute_role()
+        if not mute_role:
+            return
+        async with self.config.guild(guild).removed_users() as removed_users:
+            if str(member.id) in removed_users:
+                until_ts = removed_users[str(member.id)]
+                until = datetime.fromtimestamp(until_ts, tz=timezone.utc) if until_ts else None
+                # datetime is required to utilize the mutes method
+                if until and until < datetime.now(tz=timezone.utc):
+                    return
+                removed_users.pop(str(member.id))
+                role = guild.get_role(mute_role)
+                if not role:
+                    return
+                await self.mute_user(
+                    guild, guild.me, member, until, _("Previously muted in this server.")
+                )
+
+    @commands.group()
+    @commands.guild_only()
+    @checks.mod_or_permissions(manage_roles=True)
```

Comment:
This makes more sense to be limited to bot's admin role because it involves role-related functionality as well as changes other cog's settings that affect the whole guild.
```suggestion
    @checks.admin_or_permissions(manage_roles=True)
```